staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,640 @@
+func (qs *QueueSetImpl) updateQueueVirStartTime(packet *Packet, queue *Queue) {
+	// When a request arrives to an empty queue with no requests executing:
+	// len(queue.GetPackets()) == 1 as enqueue has just happened prior (vs  == 0)
+	if len(queue.GetPackets()) == 1 && queue.GetRequestsExecuting() == 0 {
+		// the queue’s virtual start time is set to the virtual time.
+		queue.SetVirStart(qs.vt)
+	}
+}
+
+// removeQueueAndUpdateIndexes uses reslicing to remove an index from a slice
+//  and then updates the 'Index' field of the queues to be correct
+func removeQueueAndUpdateIndexes(queues []*Queue, index int) []*Queue {
+	removedQueues := removeIndex(queues, index)
+	for i := index; i < len(removedQueues); i++ {
+		removedQueues[i].SetIndex(removedQueues[i].GetIndex() - 1)
+	}
+	return removedQueues
+}
+
+// removeIndex uses reslicing to remove an index from a slice
+func removeIndex(s []*Queue, index int) []*Queue {
+	return append(s[:index], s[index+1:]...)
+}
+
+// FinishPacketAndDequeueWithChannelAsMuchAsPossible is a convenience method which calls finishPacket
+// for a given packet and then dequeues as many packets as possible
+// and updates that packet's channel signifying it is is dequeued
+// this is a callback used for the filter that the QueueSetImpl supports
+func (qs *QueueSetImpl) FinishPacketAndDequeueWithChannelAsMuchAsPossible(pkt *Packet) {
+	qs.LockAndSyncTime()
+	defer qs.lock.Unlock()
+
+	qs.finishPacket(pkt)
+	qs.DequeueWithChannelAsMuchAsPossible()
+}
+
+// FinishPacket is a callback that should be used when a previously dequeud packet
+// has completed it's service.  This callback updates imporatnt state in the
+// QueueSetImpl
+func (qs *QueueSetImpl) finishPacket(p *Packet) {
+
+	S := qs.clk.Since(p.GetStartTime()).Seconds()
+
+	// When a request finishes being served, and the actual service time was S,
+	// the queue’s virtual start time is decremented by G - S.
+	virstart := p.GetQueue().GetVirStart()
+	virstart -= qs.estimatedServiceTime - S
+	p.GetQueue().SetVirStart(virstart)
+
+	// request has finished, remove from requests executing
+	requestsExecuting := p.GetQueue().GetRequestsExecuting()
+	requestsExecuting--
+	p.GetQueue().SetRequestsExecuting(requestsExecuting)
+
+	// Logic to remove quiesced queues
+	// >= as QueueIdx=25 is out of bounds for desiredNumQueues=25 [0...24]
+	if p.GetQueue().GetIndex() >= qs.desiredNumQueues &&
+		len(p.GetQueue().GetPackets()) == 0 &&
+		p.GetQueue().GetRequestsExecuting() == 0 {
+		qs.queues = removeQueueAndUpdateIndexes(qs.queues, p.GetQueue().GetIndex())
+		// At this point, if the qs is quiescing,
+		// has zero requests executing, and has zero requests enqueued
+		// then a call to the EmptyHandler should be forked.
+		if qs.quiescent && qs.numPacketsEnqueued == 0 &&
+			qs.GetRequestsExecuting() == 0 {
+			// then a call to the EmptyHandler should be forked.
+			go func() {
+				// TODO(aaron-prindle) store the emptyHandler to call it here?
```

Comment:
yes.  Or, more simply, `go qs.emptyHandler()`.

Reply:
updated to `go qs.emptyHandler.HandleEmpty()`